### ISSUE TITLE
Reduce rendering to 1-30 fps on Mac OS X will solve overheat problem?

### ISSUE THREAD

**Issue by wildfiremedia** [photonstorm/phaser#1968]
Mac OS X has been a teething issue with heat problem that can't be cool with internal fans, even at 6200 RPM max speed, we still getting up to 100'C temperature and power consumption on Intel will reach 24 Watts at peak load. Idle CPU load was less than 1 Watt and 45'C. The temperature and CPU load was obtained through Intel Power Gadget tool. The best controlled temperature on Macbook Pro is 5 Watts+ @ 50'C as possible on Windows with energy saving mode in Windows power management.

I propose Phaser could be optimal with an option to lowering framerate to 1-30 fps for this use case where Mac OS X has no way to limit CPU frequency internally or any 3rd party tools and would require a fix with urgency.

**Comment by dotContent** [comment]
Hi! I've been spending a bunch of time with this so I thought I'd step in and comment.

Phaser's target FPS is 60 - it is not adjustable. But there are various things you can do to reduce your framerate or target framerate.

You seem to be able to adjust a targetFPS at game.time.desiredFps. This doesn't appear to adjust the actual framerate, so I'm assuming what it does is set the number of times it does the calculations to determine the state to render. I found that anything below 10 caused issues with interactivity for me.

You can also disable rendering new frames entirely. Someone used this option to create an FPS limiter plugin at https://github.com/photonstorm/phaser-plugins/tree/master/SaveCPU , however it doesn't work with the newest version of Phaser. See https://github.com/photonstorm/phaser-plugins/issues/10 for a hastily made half solution to getting it work with the newest version. Your milage may also vary with this one as well - I found that the framerate was much lower than it should have been. However, it did reduce CPU usage.

However, I would not integrate either of these until you are sure you are using phaser in a way that plays nice. I particularly found that using Graphics without caching them as textures really hurt my CPU usage. I've also seen things about scaling up your game without antialiasing, not using or reducing text on the screen, reducing the number of objects your are rendering, and not using debug text at all to increase performance and reduce CPU usage. Use https://github.com/englercj/phaser-debug to get an idea of at what step your game is wasting the most CPU on.

Also, try different browsers. Each one handles both Canvas and OpenGL differently.

Either way, I'm assuming what you mean by a "heat problem that can't be cool[ed] with internal fans" is that you crash. Your hardware should never crash from overheating, especially laptops. If you are not getting a crash then you aren't working outside of the specs that the hardware is designed for OR you are having driver or hardware issues. In both cases I suggest taking your computer in to be tested to make sure it is functioning properly. Some advice: when you first get your computer, try burning it in with stress tests and record the temperatures. If your peak temps differ from those initial ones at some point during your computer's life then something changed and you may need to get your computer checked out.

Sorry for the long reply!

TL;DR: Can't change the target FPS, but there are things you can do to try to get better performance, and if you are overheating or you think your computer is getting too hot then you should take it in to get checked out.

If anyone has any more advice on reducing CPU usage and increasing Phaser performance, please chime in! Also, if any of the advice I am giving isn't correct please call me out!

**Comment by wildfiremedia** [comment]
Hi thanks for the long reply, I have narrowed down to the problem with
Macbook cooling design that is not effective at keeping low temperature
unless the chassis or cover is left open for more ventilation and CPU turbo
boost itself could be disable to prevent unwanted overheat. Having working
as a laptop and PC repair, many laptops are facing the same issues with
high temperature due to Intel i-cores bad design vs Atom or ARM, there is
really little we can do unless Apple has user define power management like
in Windows have it for a long time.

**Comment by photonstorm** [comment]
I run a Macbook Pro all day without it overheating, so perhaps this is model specific. I only use Chrome though, not Safari.

Phaser runs using request animation frame - this sets the redraw rate. What you could do is force it to use setTimeout instead (there is a game config property to do this) and then the fps rate you set will directly influence the quantity of times render is called, which may help.

**Comment by dotContent** [comment]
Thanks for the idea @photonstorm! I didn't think to try setTimeout to set the FPS!

**Comment by wildfiremedia** [comment]
@dotContent Not using any browsers but Electron that consumption lot of watts, e.g. 22w vs 11w. Passing force-gpu-rasterization parameter has improve on web browser using Phaser examples but see no changes on Electron.

Here, I passed the Google chrome parameter in Electron config if you have any idea how to setup for a test.
app.commandLine.appendSwitch("force-gpu-rasterization")
app.commandLine.appendSwitch("show-fps-counter")  //shown 60fps most of the time

@photonstorm I guess you could try loading a basic stage on Electron on Mac OS X which tend to push too hard.

**Comment by bdaenen** [comment]
@wildfiremedia Also, try using the canvas renderer for OSX if it's an option. My prototypes ran horrible on my macbook at work, but really smooth on my PC. Turns out WebGL support on most OSX/Linux platforms is just really horrible in terms of performance (and thus straining your hardware more -> more heat). This performance issue exists on the browser implementation level, not in Phaser.

I usually run some ugly regex on ``` window.navigator.userAgent ``` to check if we're on osx and force canvas if true.

**Comment by wildfiremedia** [comment]
I have found Mac OS X laptop is bad at air ventilation that give poor
cooling, same for other ultrabook on Windows. If we drill some holes like
Lenovo Yoga have, it will cool significant.

Alone disable turbo boost frequency with some 3rd party tool can help
reduce to 104'C temperature on iMovie video processing.

There is other canvas framework which is simple that Phased that ran on
Google chrome will not overheat so much. Mobile devices is a good idea
after all.

**Comment by photonstorm** [comment]
Closing this off as the discussion appears to be over, with several options presented in the thread.
